Implement an input event specific system call for SDL

alanjian85 opened the pull request:
This PR added a dedicated event system for game demo input, `syscall_poll_event` requires the user code pass in an `event_t` structure to populate, and it returns a status flag that indicates whether the event queue is not empty. The `event_t` structure is a simplified version of `SDL_Event`, it has a `type` field which may contains: `KEY_EVENT`, `MOUSE_MOTION_EVENT` and `MOUSE_BUTTON_EVENT`, these types of event are enough for the current game demo, and every event type has an associated value buffer stored in `union`, the user code may check the detail of event by accessing the corresponding value buffer.

Resolve #27

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,75 @@
 #include "state.h"
 
+#define MAX_EVENT_QUEUE_SIZE 100
+#define SCREEN_SCALE 4
```
Don't set `SCREEN_SCALE`, which is not always fitting.

alanjian85 replied:
Should I just remove the screen upsampling feature or add an extra screen_scale parameter in the `syscall_draw_frame`(and `check_sdl`), or implement the screen upsampling in the game itself? The original 640x384 resolution of quake-embedded is too high and will cause the game become neither interactive nor playable, however, 320x200 is quite efficient, but it makes the window of the game too small in my opinion.

jserv replied:
The customization and/or optimization for Quake should be landed in another pull request. Since this pull request is intended for adding input specific system calls, no functional changes against `syscall_draw_frame` would appear.

jserv replied:
Sooner or later, we can add some hotkeys which dynamically scale SDL window at runtime though.

jserv reviewed (changes requested):
Add a new file `docs/syscall.md`, which describes the SDL oriented system calls.
The Quake ELF file should be updated in another pull request. Let's concentrate on code reviewing first.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,75 @@
+
+static void event_push(event_t event)
+{
+    if (!event_queue.full) {
```
Rewrite as following:
```c
if (event_queue.full)
    return;

event_queue.events[event_queue.end++] = event;
...

```

jserv commented:
Read [How to Write a Git Commit Message](https://cbea.ms/git-commit/) carefully and improve the git commit messages.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,73 @@
 
 #include "state.h"
 
+#define MAX_EVENT_QUEUE_SIZE 100
```
Is `100` a reasonably fair number?

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,73 @@
+} key_event_t;
+
+typedef struct {
+    int32_t xrel;
```
Rewrite as following:
```c
    int32_t xrel, yrel;
```

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,73 @@
+
+typedef struct {
+    event_t events[MAX_EVENT_QUEUE_SIZE];
+    size_t start;
```
Rewrite as `size_t start, end` for short.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,73 @@
+    if (event_queue.start == event_queue.end)
+        return false;
+    *event = event_queue.events[event_queue.start++];
+    event_queue.start %= MAX_EVENT_QUEUE_SIZE;
```
If `MAX_EVENT_QUEUE_SIZE` is a power of 2, we can apply optimization technique which converts to `event_queue.start &= MAX_EVENT_QUEUE_SIZE -1`, resulting in faster execution.

jserv replied:
See https://stackoverflow.com/questions/10527581/why-must-a-ring-buffer-size-be-a-power-of-2

jserv reviewed (changes requested):
Use `git rebase -i` to squash and write down informative git commit messages.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,71 @@
 
 #include "state.h"
 
+#define MAX_EVENT_QUEUE_SIZE 128
```
Add comment for `MAX_EVENT_QUEUE_SIZE` to address its requirement: power of 2.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,17 @@
+Poll a input-specific event from the event system, `event` should contains a 32-bit `type` field, and associated with a suitably sized value buffer, the internal event queue will be updated whenever `syscall_draw_frame` or `syscall_draw_frame_pal` is called.
+
+The following types of event are currently supported:
+* `KEY_EVENT`(`0x0`): Triggered when the status of keyboard is changed, either a key is pressed or released, returns a 32-bit keycode and a 8-bit key state, the hexadecimal keycode values are listed here: https://wiki.libsdl.org/SDLKeycodeLookup
```
Replace "here: ..." with `in [SDL Keycode Lookup Table](https://wiki.libsdl.org/SDLKeycodeLookup)`.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,17 @@
+# System Calls
+
+## SDL-Oriented System Calls
```
Replace "SDL-Oriented System Calls" with "Experimental Display and Event System Calls" and mention SDL at the beginning of this paragraph.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,19 @@
+
+These system calls are experimental for demonstration of RISC-V graphics applications, they don't exist in the ABI interface of POSIX and Linux, and are only for the convenience of accessing the SDL library.
+
+### `syscall_draw_frame(screen, width, height)`
```
The subsection should be `draw_frame`, and you ought to describe the syscall number and usage.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,19 @@
+### `syscall_draw_frame(screen, width, height)`
+Update the framebuffer of the given size, the window is lazy-initialized, if it does not exist when this system call is called, a new window will be created with the given width and height. This system call also poll events from SDL library, and update the internal input specific event queue if neccessary.
+
+### `syscall_draw_frame_pal(screen, pal, width, height)`
```
Ditto. Should be `draw_frame_pal`.

jserv commented on syscall_sdl.c:
```diff
@@ -43,10 +109,55 @@ static bool check_sdl(struct riscv_t *rv, uint32_t width, uint32_t height)
+                .mouse_button = {
+                    .button = event.button.button,
+                    .state = (bool) (event.button.state == SDL_PRESSED),
+                }};
```
Add a comma between `}` and another `}`.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,19 @@
+These system calls are experimental for demonstration of RISC-V graphics applications, they don't exist in the ABI interface of POSIX and Linux, and are only for the convenience of accessing the SDL library.
+
+### `syscall_draw_frame(screen, width, height)`
+Update the framebuffer of the given size, the window is lazy-initialized, if it does not exist when this system call is called, a new window will be created with the given width and height. This system call also poll events from SDL library, and update the internal input specific event queue if neccessary.
```
Correct grammar mistakes.

jserv replied:
Some services such as [QuillBot](https://quillbot.com/) can help you.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,75 @@
+    uint32_t type;
+    union {
+        key_event_t key_event;
+        mouse_motion_t mouse_motion;
```
How about wrapping mouse in another struct? That is,
```c
struct mouse {
    mouse_motion_t motion;
    mouse_button_t button;
}
```

Then, we can refer to the members with prefix `mouse`.

alanjian85 replied:
How about a nested named union instance? A single `struct` containing all mouse events can take up unnecessary space.

jserv replied:
> How about a nested named union instance? A single `struct` containing all mouse events can take up unnecessary space.

It sounds good.

jserv commented:
The proposed git commit message:
> This commit integrates a new system call, syscall_poll_event, with SDL's event system,

Instead of "integrates," we can say, "introduces," which practically explains the motivation.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,19 @@
+
+These system calls are solely for the convenience of accessing the SDL library and are only intended for the presentation of RISC-V graphics applications. They are not present in the ABI interface of POSIX or Linux.
+
+### `draw_frame(screen, width, height)` - `0xbeef`
```
Replace with following:
```
### `draw_frame` - Draw a frame around the SDL window

system call number: `0xbeef`
SYNOPSIS
void draw_frame(screen, width, height);
```

You can check Linux Programmer's Manual for reference.

jserv commented on syscall_sdl.c:
```diff
@@ -9,9 +9,77 @@
+    uint32_t type;
+    union {
+        key_event_t key_event;
+        union {
```
I am not sure if a `union` always works as expected. Can you confirm?

jserv commented on syscall_sdl.c:
```diff
@@ -43,10 +111,64 @@ static bool check_sdl(struct riscv_t *rv, uint32_t width, uint32_t height)
+        case SDL_MOUSEMOTION: {
+            event_t new_event = {
+                .type = MOUSE_MOTION_EVENT,
+                .mouse =
```
Use `memcpy` to initiate `motion` member, resulting in shorter code and a bit faster execution.

jserv reviewed (changes requested):
Fix the conflict with the latest `master` branch.

alanjian85 committed:
Implement an input event specific system call for SDL

Everytime a new frame is drawn, all new input events are pushed into the
internal event queue, this commit introduces a new system call,
poll_event, which attempts to pop events from the internal event queue
and returns input specific event information to the user code.

Supported event types:
Keyboard Event: either a key is pressed or released
Mouse motion event: the cursor is moved
Mouse button event: either a mouse button is pressed or released

Both the mouse motion and the mouse button event have a corresponding
field under the "mouse" field in the event structure to store the event
information. However, because these two types of event require entirely
different fields in the value buffer, and are processed in different
ways in the SDL library, they are considered as different events.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,34 @@
+
+## Experimental Display and Event System Calls
+
+These system calls are solely for the convenience of accessing the SDL library and are only intended for the presentation of RISC-V graphics applications. They are not present in the ABI interface of POSIX or Linux.
```
Add a hyperlink to SDL: https://www.libsdl.org/

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,34 @@
+
+**system call number**: `0xBEEF`
+
+**synopsis**: `void draw_frame(void* screen, int width, int height)`
```
Unify pointer and reference alignment style. i.e., use
```c
void draw_frame(void *screen, int width, int height)
```

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,34 @@
+
+**system call number**: `0xBABE`
+
+**synopsis**: `void draw_frame_pal(void* screen, void* pal, int width, int height)`
```
Ditto. Style.

jserv commented on docs/syscall.md:
```diff
@@ -0,0 +1,34 @@
+
+**system call number**: `0xC0DE`
+
+**synopsis**: `int poll_event(void* event)`
```
Ditto. Style.

jserv reviewed (changes requested):
Improve the writing of git commit message:
* Fix grammar mistakes.
* Explain why mouse motion and button are considered as different events from a classification perspective.
* Remove the statement " this new system will improve the interactivity of this emulator" since it is obvious.

jserv commented:
Thank @alanjian85 for contributing!
